### PR TITLE
Simplify Play-GoogleAuth integration

### PR DESCRIPTION
**Pull request by adamnfish**
This PR does two things:

1. Move the OAuth logic into the library to dramatically simplify the integration process
2. Update the example app

For the former, the main change is offering 2 versions of a `processOauth2Callback` method that handles the OAuth logic with and without Google Group enforcement. This was logic that was being copy-pasted in every project.
It also offers a `startGoogleLogin` method to kick off the process because there was some fiddly logic in there that everyone was duplicating as well.

To support this there are a couple of new requirements for the Controllers that implement the Play-GoogleAuth trait. These are providing a couple of additional routes (to specify the actions to use for login failure and a default redirect in case there's no redirect URL present in the session) and providing a WsClient instance so the library can make the OAuth requests itself.

To document these changes, the example app has been updated to use the new helpers and provide the newly-required items.
Additionally, the example app now works a lot more like a real app would. The settings are drawn from the config and the required bits are injected in using Play. This should help people work out how to *practically* integrate with Play-GoogleAuth, rather than just showing them how a toy implementation looks.

This is a pretty big (and breaking) change so I'm open to bumping the version to `0.5.0` when this is merged, rather than just going to `0.4.1`. this would also line up this library's minor version with that of Play's which would be a bit less confusing.

**NB:** I've tested the example app but ideally we could test a 'proper' integration before this gets merged and released.

**NB:** In its proposed form, this version of Play-GoogleAuth depends on Cats `0.6.0`, which is not yet a stable release. This means that if a breaking change were to be made to Cats, Weird Things™ would happen. It'd be possible to change this PR so it doesn't have a Cats dependency, I'm happy to discuss that option here.

### CODE REVIEW

**Comment by cb372** [comment]
Sorry, I forgot about this. I promise I'll look at it tomorrow!

File: module/src/main/scala/com/gu/googleauth/actions.scala [hunk 1]
```diff
@@ -41,29 +48,139 @@ trait UserIdentifier {
...
+  /**
+    * Looks up user's Google Groups and ensures they belong to any that are required. Redirects to
+    * `failureRedirectTarget` if the user is not a member of any required group.
+    */
+  def enforceGoogleGroups(userIdentity: UserIdentity, requiredGoogleGroups: Set[String], googleGroupChecker: GoogleGroupChecker)
+                         (implicit request: RequestHeader): XorT[Future, Result, Unit] = {
+    googleGroupChecker.retrieveGroupsFor(userIdentity.email).attemptT
+      .leftMap { t =>
+        Logger.warn("Login failure, Could not look up user's Google groups", t)
+        redirectWithError(failureRedirectTarget, "Unable to look up user's 2FA status", authConfig.antiForgeryKey, request.session)
```
Comment (cb372) [inline module/src/main/scala/com/gu/googleauth/actions.scala]: All these log messages and error messages related to 2FA should be rewritten to be more generic, as we are looking for an arbitrary set of required Google groups here.

File: module/src/main/scala/com/gu/googleauth/actions.scala [hunk 2]
```diff
@@ -74,6 +191,13 @@ trait Actions extends UserIdentifier {
   object AuthAction extends AuthenticatedBuilder(r => userIdentity(r), r => sendForAuth(r))
 }
 
+object Actions {
+  private[googleauth] def checkGoogleGroups(userGroups: Set[String], requiredGroups: Set[String]): Boolean = {
+    if (userGroups.intersect(requiredGroups) == requiredGroups) true
```
Comment (cb372) [inline module/src/main/scala/com/gu/googleauth/actions.scala]: Wrapping the predicate with `if () true else false` seems a bit verbose to me

**Comment by cb372** [comment]
Looks fine to me apart from the incorrect logging. I'm fine with the Cats dependency.

I agree that version 0.5.0 is sensible for this.

File: module/src/main/scala/com/gu/googleauth/actions.scala (hunk 1 above)
Comment (adamnfish) [inline module/src/main/scala/com/gu/googleauth/actions.scala]: Great spot, thanks

File: module/src/main/scala/com/gu/googleauth/actions.scala (hunk 2 above)
Comment (adamnfish) [inline module/src/main/scala/com/gu/googleauth/actions.scala]: Ooops.

**Comment by adamnfish** [comment]
Looks like `0.6.0` is out now as well :-/

I can do an update to add that afterwards as well.

**Comment by adamnfish** [comment]
This worked ok [in Amiable](https://github.com/guardian/amiable/pull/31) so I think it's (past) time to merge it.

I'll do so and release it at version `0.5.0` unless anyone has any objections.

**Pull request merged**